### PR TITLE
Make grid

### PR DESCRIPTION
**Pull request by loic425**
Replaces #46

### CODE REVIEW

**Comment by lbaey** [comment]
👍

File: src/Bundle/Maker/MakeGrid.php [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        );
+
+        $generator->writeChanges();
+
+        $this->writeSuccessMessage($io);
+    }
+
+    public function configureDependencies(DependencyBuilder $dependencies): void
+    {
+        // noop
```
Comment (lchrusciel) [inline src/Bundle/Maker/MakeGrid.php:121]: ?

File: src/Bundle/Tests/Functional/Maker/MakeGridTest.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    {
+        $tester = new CommandTester((new Application(self::bootKernel()))->find('make:grid'));
+
+        $this->assertFileDoesNotExist(self::tempFile('Grid/PriceGrid.php'));
+
+        $tester->execute(['entity' => Price::class, '--namespace' => 'Tests\Tmp\Grid']);
+
+        $this->assertFileExists(self::tempFile('Grid/PriceGrid.php'));
+        $this->assertSame(self::getPriceGridExpectedContent(), \file_get_contents(self::tempFile('Grid/PriceGrid.php'))
+        );
```
Comment (lchrusciel) [inline src/Bundle/Tests/Functional/Maker/MakeGridTest.php:34]:
```suggestion
        $this->assertSame(self::getPriceGridExpectedContent(), \file_get_contents(self::tempFile('Grid/PriceGrid.php')));
```

File: src/Bundle/Tests/Functional/Maker/MakeGridTest.php [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    /** @test */
+    public function it_can_create_grids(): void
+    {
+        $tester = new CommandTester((new Application(self::bootKernel()))->find('make:grid'));
+
+        $this->assertFileDoesNotExist(self::tempFile('Grid/PriceGrid.php'));
+
+        $tester->execute(['entity' => Price::class, '--namespace' => 'Tests\Tmp\Grid']);
+
+        $this->assertFileExists(self::tempFile('Grid/PriceGrid.php'));
```
Comment (lchrusciel) [inline src/Bundle/Tests/Functional/Maker/MakeGridTest.php:32]: Can we extract `'Grid/PriceGrid.php'` to const, so we will be sure that the path is identical?

File: src/Bundle/Resources/config/skeleton/Grid.tpl.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        echo "                    ->setLabel('".ucfirst($fieldname)."')\n";
+        echo "             )\n";
+    }
+
+    if ('BOOLEAN' === $type) {
+        echo "            ->addField(\n";
+        echo "                TwigField::create('".$fieldname."', 'path/to/field/template.html.twig')\n";
+        echo "                    ->setLabel('".ucfirst($fieldname)."')\n";
+        echo "             )\n";
+    }
```
Comment (lchrusciel) [inline src/Bundle/Resources/config/skeleton/Grid.tpl.php:58]: I'm not sure about this node. What is so special about boolean node, that it is handled with TwigField? I know, that this is how we do it mostly in Sylius, but the current behaviour will result in 500 error, as the template path does not exist. I would use a string field to be sure, that the generated grid can work just after the generation

File: .gitignore [hunk 5]
```diff
@@ -4,3 +4,4 @@
 /composer.lock
 
 /.phpunit.result.cache
+src/Bundle/test/tmp
```
[the remark this answers is not in the document]
Comment (loic425) [reply .gitignore:7]: The directory is automatically created by Symfony maker command (I guess using symfony/filesystem), so it is not necessary.

File: src/Bundle/Maker/MakeGrid.php (hunk 1 above)
Comment (loic425) [reply src/Bundle/Maker/MakeGrid.php:121]: It's required by the interface `Symfony\Bundle\MakerBundle\MakerInterface'.
But I have to check the return typehint on Symfony 6.

File: src/Bundle/Resources/config/skeleton/Grid.tpl.php (hunk 4 above)
Comment (loic425) [reply src/Bundle/Resources/config/skeleton/Grid.tpl.php:58]: Yes you're right, we need a generated grid which work just after the genration. I agree.

Maybe I should comment this code lines just to show you can handle it with a twig template.
and maybe add a comment with the "@SyliusUi/Grid/Field/enabled.html" for Sylius/Monofony users too...

Comment (loic425) [reply src/Bundle/Resources/config/skeleton/Grid.tpl.php:58]: WDYT?

File: src/Bundle/Tests/Functional/Maker/MakeGridTest.php (hunk 3 above)
Comment (loic425) [reply src/Bundle/Tests/Functional/Maker/MakeGridTest.php:32]: Good idea, on this class or in MarkerTestCase?

File: src/Bundle/Maker/MakeGrid.php (hunk 1 above)
Comment (loic425) [reply src/Bundle/Maker/MakeGrid.php:121]: No that's good, that's not part of Symfony monorepository
https://github.com/symfony/maker-bundle/blob/main/src/MakerInterface.php

Comment (loic425) [reply src/Bundle/Maker/MakeGrid.php:121]: https://github.com/symfony/maker-bundle/blob/006577909ded549ee48fd69aa4da1a75ff9a2b55/src/Maker/MakeSerializerNormalizer.php

File: src/Bundle/Resources/config/skeleton/Grid.tpl.php (hunk 4 above)
Comment (mbabker) [reply src/Bundle/Resources/config/skeleton/Grid.tpl.php:58]: Better question.  Why doesn't this bundle provide a `BooleanField` that's usable out-of-the-box?  A boolean field is a pretty common use case, it shouldn't need a Twig field to handle rendering unless you're aiming to do some custom display logic that a "simple" configuration couldn't handle ("simple" as in options to control the output text for both states, if you want to start using badges or icons for output that's where a Twig field is better suited).

Comment (loic425) [reply src/Bundle/Resources/config/skeleton/Grid.tpl.php:58]: That's a good reflexion, not for this PR, but for the future of this bundle.

Comment (loic425) [reply src/Bundle/Resources/config/skeleton/Grid.tpl.php]: But for sure, with TwigField (if we use it for other Doctrine types) have to be code-commented to keep the grid working just after the generation.

File: src/Bundle/Tests/Functional/Maker/MakeGridTest.php (hunk 3 above)
Comment (loic425) [reply src/Bundle/Tests/Functional/Maker/MakeGridTest.php]: Done on this class, cause it's grid related. If we add some other makers on this bundle, it shouldn't be related.

File: src/Bundle/Maker/MakeGrid.php (hunk 1 above)
Comment (lchrusciel) [reply src/Bundle/Maker/MakeGrid.php:121]: I meant - what do you mean by noop? No operation? I have the feeling, that both of us should be more specific :) I with my question and you with the comment here, so people would not need to try to resolve what is `noop`

File: src/Bundle/Tests/Functional/Maker/MakeGridTest.php [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+use App\Entity\Book;
+use App\Entity\Price;
+use Symfony\Bundle\FrameworkBundle\Console\Application;
+use Symfony\Bundle\MakerBundle\Exception\RuntimeCommandException;
+use Symfony\Component\Console\Tester\CommandTester;
+
+final class MakeGridTest extends MakerTestCase
+{
+    public const BOOK_GRID_PATH = 'Grid/BookGrid.php';
```
Comment (lchrusciel) [inline src/Bundle/Tests/Functional/Maker/MakeGridTest.php:24]: Do we need them to be public? 
```suggestion
    private const BOOK_GRID_PATH = 'Grid/BookGrid.php';
```

File: src/Bundle/DependencyInjection/SyliusGridExtension.php [hunk 7]
```diff
@@ -57,6 +59,10 @@ public function load(array $configs, ContainerBuilder $container): void
         $container->registerForAutoconfiguration(GridInterface::class)
             ->addTag('sylius.grid')
         ;
+
+        if (!\class_exists(AbstractMaker::class)) {
```
Comment (lchrusciel) [inline src/Bundle/DependencyInjection/SyliusGridExtension.php:63]: Can we somehow write test case for it? I'm not sure. The best idea I have right now is to duplicate test app without maker bundle... Did you test it manually?

File: src/Bundle/Maker/MakeGrid.php (hunk 1 above)
Comment (loic425) [reply src/Bundle/Maker/MakeGrid.php:121]: I don't know lol
That's just copied/paste from https://github.com/zenstruck/foundry/blob/master/src/Bundle/Maker/MakeFactory.php#L170

File: src/Bundle/Tests/Functional/Maker/MakeGridTest.php (hunk 6 above)
Comment (loic425) [reply src/Bundle/Tests/Functional/Maker/MakeGridTest.php]: I made that change on the three constants.

File: src/Bundle/DependencyInjection/SyliusGridExtension.php (hunk 7 above)
Comment (loic425) [reply src/Bundle/DependencyInjection/SyliusGridExtension.php:63]: I didn't test it manually, but I have an idea....

Comment (loic425) [reply src/Bundle/DependencyInjection/SyliusGridExtension.php]: Voilà!
https://github.com/Sylius/SyliusGridBundle/pull/206/commits/b76ec9fcfab96d32fde3d752b08851231124e3b8

Comment (loic425) [reply src/Bundle/DependencyInjection/SyliusGridExtension.php]: done... it wasn't as simple as I thought 🙄

**Review by lchrusciel — APPROVED** [review]

File: src/Bundle/Resources/config/skeleton/Grid.tpl.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+}
+?>
+            ->addActionGroup(
+                MainActionGroup::create(
+                    CreateAction::create(),
+                )
+            )
+            ->addActionGroup(
+                ItemActionGroup::create(
+                    // ShowAction::create(),
```
Comment (lchrusciel) [inline src/Bundle/Resources/config/skeleton/Grid.tpl.php:68]: I would at least leave the comment why do we have it commented out

File: src/Bundle/Resources/config/skeleton/Grid.tpl.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+    if (str_starts_with($type, 'DATE')) {
+        echo "            ->addField(\n";
+        echo "                DateTimeField::create('".$fieldname."')\n";
+        echo "                    ->setLabel('".ucfirst($fieldname)."')\n";
+        echo "             )\n";
+    }
+
+    if ('BOOLEAN' === $type) {
+        echo "             //->addField(\n";
```
Comment (lchrusciel) [inline src/Bundle/Resources/config/skeleton/Grid.tpl.php:54]: I would at least leave the comment why do we have it commented out

**Review by lchrusciel — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thank you, Loïc! :tada:

**Pull request merged**